### PR TITLE
tls/x509utils/certpool: cached Export() and non-sharing Copy()

### PR DESCRIPTION
**Pull request by amery**
## Summary by CodeRabbit

- **New Features**
	- Enhanced `CertPool` structure to include a caching mechanism for improved performance.
	- New methods for cloning and validating certificate entries.

- **Bug Fixes**
	- Improved error handling and logic in certificate import and deletion processes.
	- Ensured only valid certificates are exported.

- **Refactor**
	- Streamlined methods for better readability and efficiency, including updates to the `Delete` and `Import` methods.
	- Updated logic in the `Export` and `Copy` methods to utilize the cache effectively.

### CODE REVIEW

**Comment by amery** [comment]
amended

File: tls/x509utils/certpool/certpool_copy.go [hunk 1]
```diff
@@ -9,18 +9,41 @@ import (
...
+	out := x509.NewCertPool()
+	for _, ce := range s.hashed {
+		out.AddCert(ce.cert)
 	}
 	return out
 }
 
+func (s *CertPool) unsafeInvalidateCache() {
+	s.cache = nil
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/certpool/certpool_copy.go:45]: that's why it's called _unsafe_ ;-)

File: tls/x509utils/certpool/certpool_copy.go [hunk 2]
```diff
@@ -9,18 +9,41 @@ import (
...
 
-		for _, ce := range s.hashed {
-			out.AddCert(ce.cert)
-		}
+	s.mu.RLock()
+	if out := s.cache; out != nil {
+		// cached
+		s.mu.RUnlock()
+		return out
+	}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/certpool/certpool_copy.go:21]: for SystemCertPool() you were advocating the use for RWLock ... well

File: tls/x509utils/certpool/certpool_copy.go [hunk 3]
```diff
@@ -9,18 +9,37 @@ import (
...
-		defer s.mu.RUnlock()
+	if s == nil {
+		return x509.NewCertPool()
+	}
 
-		for _, ce := range s.hashed {
-			out.AddCert(ce.cert)
-		}
+	s.mu.Lock()
+	defer s.mu.RUnlock()
```
Comment (amery) [inline tls/x509utils/certpool/certpool_copy.go:17]: oops

File: tls/x509utils/certpool/certpool_copy.go [hunk 4]
```diff
@@ -9,18 +9,33 @@ import (
...
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	out := s.cache
+	if out == nil {
+		out = s.unsafeExport()
+		s.cache = out
+	}
+	return out
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/certpool/certpool_copy.go]: @coderabbit commit amended following your latest suggestion. please review again

File: tls/x509utils/certpool/certpool_writer.go [hunk 5]
```diff
@@ -49,37 +49,25 @@ func (s *CertPool) Delete(ctx context.Context, name string) error {
...
-		// entries to unique hashes
-		core.SliceUniquify(&entries)
-		hashes := make([]Hash, len(entries))
-		for i, ce := range entries {
-			hashes[i] = ce.hash
-		}
+	// convert name to unique hashes
+	hashes := core.SliceAsFn(func(ce *certPoolEntry) (Hash, bool) {
+		return ce.Hash()
+	}, s.getListForName(sn).Values())
```
[the remark this answers is not in the document]
Comment (amery) [reply tls/x509utils/certpool/certpool_writer.go:62]: check ce.Hash()'s signature

**Review by karasz — APPROVED** [review]
LGTM

**Pull request merged**